### PR TITLE
Add Phenyl API explorer

### LINKED ISSUE #18: Create module: "phenyl-api-explorer"
## What's it?
"phenyl-api-explorer" automatically renders API explorer pages like swagger by being given `FunctionalGroup`.

## Provisional usage

```js
import http from 'http'
import PhenylApiExplorer from 'phenyl-api-explorer/jsnext'
import PhenylHttpServer from 'phenyl-http-server/jsnext'
import PhenylCore from 'phenyl-core/jsnext'

const functionalGroup = {} // Here you build your original one

const server = new PhenylHttpServer(http.createServer(), {
  runner: new PhenylCore(functionalGroup),
  customRequestHandler: new PhenylApiExplorer({ path: '/explorer' }, functionalGroup).handler
})

server.listen(8080)
```

Then, you can get API explorer page by `http://localhost:8080/explorer`.

### PR DESCRIPTION
**Pull request by Leko**
Closes #18

## How to develop

```
cd modules/phenyl-api-explorer
cd src/client
yarn && yarn build
cd ../../
npm start
open http://localhost:8000/explorer
```

### CODE REVIEW

File: modules/phenyl-api-explorer/src/PhenylApiExplorer.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  html: string;
+
+  constructor (functionalGroup: FunctionalGroup, params: ExplorerParams) {
+    this.functionalGroup = functionalGroup
+    this.path = params.path
+    this.handler = this.handler.bind(this)
+    this.html = this.getClientHTML()
+  }
+
+  shallowMap (obj: Object, fn: (any, string, Object) => any): Object {
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/PhenylApiExplorer.js]: Is this just a utility function? If so, why don't you separate it into a plain function?

File: modules/phenyl-api-explorer/src/PhenylApiExplorer.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    return ret
+  }
+
+  getClientHTML (): string {
+    const templatePath = path.join(__dirname, 'client', 'build', 'index.html')
+    const template = fs.readFileSync(templatePath, 'utf8')
+    const data = {
+      functionalGroup: {
+        users: this.shallowMap(this.functionalGroup.users, ({ accountPropName, passwordPropName }) => {
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/PhenylApiExplorer.js]: This depends on standard definition. We should discuss how to pass extra data to the renderer.

File: modules/phenyl-api-explorer/src/PhenylApiExplorer.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    for (let p in obj) {
+      ret[p] = fn(obj[p], p, obj)
+    }
+    return ret
+  }
+
+  getClientHTML (): string {
+    const templatePath = path.join(__dirname, 'client', 'build', 'index.html')
+    const template = fs.readFileSync(templatePath, 'utf8')
+    const data = {
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/PhenylApiExplorer.js]: Define the type?

```js
type FunctionalGroupSkeleton = {
   nonUsers: {[string]: true},
...
}
```

File: modules/phenyl-api-explorer/src/client/config/env.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+const NODE_ENV = process.env.NODE_ENV
+if (!NODE_ENV) {
+  throw new Error(
+    'The NODE_ENV environment variable is required but was not specified.'
+  )
+}
+
+// https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use
+var dotenvFiles = [
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/client/config/env.js]: `const` insteadof `var` ?

File: modules/phenyl-api-explorer/src/client/config/env.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
+/* eslint-env node */
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/client/config/env.js]: What's the main reason you didn't use flow or ES module import / export here?

File: modules/phenyl-api-explorer/src/client/src/App.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
+import React, { Component } from 'react'
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/client/src/App.js]: No flow pragma here?

File: modules/phenyl-api-explorer/src/client/src/components/OperationEditor/index.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
+import React, { Component } from 'react'
+import { Form } from 'semantic-ui-react'
+
+type Props = {
+  match: any,
+  sessionId: string,
+  operations: Array<string>,
+  defaultPayloads: { [string]: Object },
+  execute: ({ entitiName: string, operation: string, payload: any }) => any,
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/client/src/components/OperationEditor/index.js]: `entityName`

File: modules/phenyl-api-explorer/src/client/src/components/OperationEditor/index.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  handleChangeOperation = (event, { value }) => {
+    if (!this.props.defaultPayloads[value]) {
+      throw new Error(`Unknown operation: ${value}`)
+    }
+
+    const payload = this.props.defaultPayloads[value]
+
+    this.setState({
+      operation: value,
+      payload: JSON.stringify(payload, null, 2),
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/client/src/components/OperationEditor/index.js]: You use payload as string here, but the type definition is Object.

File: modules/phenyl-api-explorer/src/client/src/components/OperationEditor/index.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  handleRun = () => {
+    const { match, execute } = this.props
+
+    const payload = JSON.parse(this.state.payload)
+    execute({
+      sessionId: this.props.sessionId,
+      entityName: match.params.functional,
+      operation: this.state.operation,
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/client/src/components/OperationEditor/index.js]: I prefer `method` over `operation`, as operation has different meanings in phenyl world.
Would you mind if you change the name?

File: modules/phenyl-api-explorer/src/client/scripts/prepare-assets.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  $('<script></script>').text(content).appendTo('body')
+
+  debug(`Remove ${assetPath}`)
+  cheerio(el).remove()     // Remove from HTML
+  fs.unlinkSync(assetPath) // Remove file
+})
+
+// Inject entities
+const entitiesDefinition = cheerio('<script></script>').text(`
+  window.EntitiesDefinition = <%- JSON.stringify(functionalGroup) %>
```
Comment (shinout) [inline modules/phenyl-api-explorer/src/client/scripts/prepare-assets.js]: OK, this name will be PhenylFunctionalGroupSkeleton.
I think global variable should contain specific `Phenyl` prefix.

File: modules/phenyl-api-explorer/src/PhenylApiExplorer.js (hunk 2 above)
Comment (shinout) [reply modules/phenyl-api-explorer/src/PhenylApiExplorer.js]: Let's throw an error indicating that `passwordPropName` and `accountPropName` are required properties.

File: modules/phenyl-api-explorer/src/client/config/env.js (hunk 5 above)
Comment (Leko) [reply modules/phenyl-api-explorer/src/client/config/env.js]: This file is copy of create-react-app.

**Comment by shinout** [comment]
Thanks @Leko for this great function!

**Pull request merged**